src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -224,22 +226,68 @@ public override TaintedDataAbstractValue VisitObjectCreation(IObjectCreationOper
                     ProcessTaintedDataEnteringInvocationOrCreation(method, taintedArguments, originalOperation);
                 }
 
-                if (this.IsSanitizingMethod(method))
+                PooledHashSet<IsInvocationTaintedWithPointsToAnalysis> evaluateWithPointsToAnalysis = null;
+                PooledHashSet<IsInvocationTaintedWithValueContentAnalysis> evaluateWithValueContentAnalysis = null;
+                try
                 {
-                    return TaintedDataAbstractValue.NotTainted;
-                }
-                else if (this.DataFlowAnalysisContext.SourceInfos.IsSourceMethod(method))
-                {
-                    return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                    if (this.IsSanitizingMethod(method))
+                    {
+                        return TaintedDataAbstractValue.NotTainted;
+                    }
+                    else if (this.DataFlowAnalysisContext.SourceInfos.IsSourceMethod(
+                        method,
+                        out evaluateWithPointsToAnalysis,
+                        out evaluateWithValueContentAnalysis))
+                    {
+                        ImmutableArray<IArgumentOperation> argumentOperation = (originalOperation as IInvocationOperation).Arguments;
+                        if (evaluateWithPointsToAnalysis != null)
+                        {
+                            IEnumerable<PointsToAbstractValue> pointsToAnalysisResultOpt = argumentOperation.Select(o => GetPointsToAbstractValue(o.Value));
+                            if (pointsToAnalysisResultOpt == null
+                                || !evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResultOpt)))
+                            {
+                                return TaintedDataAbstractValue.NotTainted;
+                            }
+                        }
+                        else if (evaluateWithValueContentAnalysis != null)
+                        {
+                            IEnumerable<PointsToAbstractValue> pointsToAnalysisResultOpt = argumentOperation.Select(o => GetPointsToAbstractValue(o.Value));
+                            IEnumerable<ValueContentAbstractValue> valueContentAnalysisResultOpt = argumentOperation.Select(o => GetValueContentAbstractValue(o.Value));
+                            if (valueContentAnalysisResultOpt == null
+                                || pointsToAnalysisResultOpt == null
+                                || !evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResultOpt, valueContentAnalysisResultOpt)))
+                            {
+                                return TaintedDataAbstractValue.NotTainted;
+                            }
+                        }
```

Comment:
1. Since evaluateWithPointsToAnalysis and evaluateWithValueContentAnalysis aren't mutually exclusive, we shouldn't have an else if.
2. Checking `!evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResultOpt)` and `!evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResultOpt, valueContentAnalysisResultOpt)` seems backwards.  I think in your other comments, you're saying evaluation returning true is tainted?  That makes sense to me.

We should instead of check:

```csharp
if ((evaluateWithPointsToAnalysis != null && evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResultOpt))
    || (evaluateWithValueContentAnalysis != null && evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResultOpt, valueContentAnalysisResultOpt))
{
    return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
}
```

Reply:
Yes, meet any one is tainted.
It's supposed to be both OK no matter which check go first, in our case:
(1) Neither evaluateWithPointsToAnalysis nor evaluateWithValueContentAnalysis is null, it will do PointsTo check first, if it doesn't meet the requirement, go to check ValueContent.
(2) evaluateWithPointsToAnalysis is not null while evaluateWithValueContentAnalysis is, so just do the PointsTo check.
(3) Both is null, IsSourceMethod() will return false, this block won't be executed.